File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.ts
function topKFrequent(nums: number[], k: number): number[] {
    const freq = new Map<number, number>();

    for (const num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1);
    }

    const bucket: number[][] = []
    for (let i = 0; i < nums.length + 1; i++) {
        bucket.push([])
    }

    for (const [num, count] of freq) {
        bucket[count].push(num);
    }

    const ans: number[] = []

    for (let i = bucket.length-1; i >= 0 && ans.length < k; i--) {
        for (const num of bucket[i]) {
            ans.push(num)
            if (ans.length === k) return ans
        }
    }
    return ans
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = {}

        for num in nums:
            counts[num] = counts.get(num, 0) + 1

        heap = [(-value, key) for key, value in counts.items()]
        heapq.heapify(heap)

        ans = []
        while len(ans) < k:
            ans.append(heapq.heappop(heap)[1])

        return ans